src/fmt.rs

```diff
@@ -0,0 +1,210 @@
+use opentelemetry::trace::TraceContextExt;
+use serde::{
+    Deserialize, Deserializer as _, Serialize, Serializer as _,
+    de::{Error, MapAccess, Visitor as DeVisitor},
+    ser::{SerializeMap, SerializeSeq},
+};
+use serde_json::{Deserializer, Serializer, Value};
+use std::{fmt, io, marker::PhantomData, ops::Deref, str};
+use tracing::{Event, Span, Subscriber};
+use tracing_opentelemetry::OpenTelemetrySpanExt;
+use tracing_serde::{AsSerde, SerdeMapVisitor};
+use tracing_subscriber::{
+    fmt::{
+        FmtContext, FormatEvent, FormatFields, FormattedFields,
+        format::Writer,
+        time::{FormatTime, SystemTime},
+    },
+    registry::{LookupSpan, SpanRef},
+};
+
+pub struct JsonFormat;
+
+impl<S, N> FormatEvent<S, N> for JsonFormat
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    fn format_event(
+        &self,
+        ctx: &FmtContext<'_, S, N>,
+        mut writer: Writer<'_>,
+        event: &Event<'_>,
+    ) -> fmt::Result {
+        let mut timestamp = String::new();
+        SystemTime.format_time(&mut Writer::new(&mut timestamp))?;
+
+        let meta = event.metadata();
+
+        let mut visit = || {
+            let mut serializer = Serializer::new(IoWriter(&mut writer));
+            let mut serializer = serializer.serialize_map(None)?;
+
+            serializer.serialize_entry("timestamp", &timestamp)?;
+            serializer.serialize_entry("level", &meta.level().as_serde())?;
+
+            // add all event fields to the json object
+            let mut visitor = SerdeMapVisitor::new(serializer);
+            event.record(&mut visitor);
+            serializer = visitor.take_serializer()?;
+
+            serializer.serialize_entry("target", meta.target())?;
+
+            // extract tracing information from the current span context
+            let current_span = Span::current();
+            if let Some(id) = current_span.id() {
+                let otel_ctx = current_span.context();
+                let span_ref = otel_ctx.span();
+                let span_context = span_ref.span_context();
+
+                if let Some(leaf_span) = ctx.span(&id).or_else(|| ctx.lookup_current()) {
+                    let spans = SpanScope(leaf_span, PhantomData::<N>);
+                    serializer.serialize_entry("spans", &spans)?;
+                }
+
+                let trace_id = span_context.trace_id().to_string();
+                serializer.serialize_entry("trace_id", &trace_id)?;
+
+                let span_id = span_context.span_id().to_string();
+                serializer.serialize_entry("span_id", &span_id)?;
+            }
+
+            SerializeMap::end(serializer)
+        };
+
+        visit().map_err(|_| fmt::Error)?;
+        writeln!(writer)
+    }
+}
+
+struct IoWriter<'a>(&'a mut dyn fmt::Write);
```

Comment:
`IoWriter` is a copy-pasted `WriteAdaptor` from `tracing_subscriber::fmt::writer`.

`SpanData` and `SpanScope` are sightly modified and slightly optimised versions of `SerializableSpan` and `SerializableContext` from `tracing_subscriber::fmt::format::json`.

I had to copy-paste those since they are not public.

I found an issue for `tracing-subscriber` asking to make those things public from 2020: https://github.com/tokio-rs/tracing/issues/1039